nx/lib/nx.ex

```diff
@@ -3546,11 +3585,179 @@ defmodule Nx do
   @doc type: :backend
   def backend_deallocate(tensor_or_container) do
     Nx.Defn.Composite.reduce(tensor_or_container, :ok, fn
-      %Nx.Tensor{} = tensor, :ok -> impl!(tensor).backend_deallocate(tensor)
-      _, :ok -> :ok
+      %Nx.Tensor{} = tensor, :ok ->
+        Nx.Shared.raise_vectorized_not_implemented_yet(tensor, __ENV__)
+        impl!(tensor).backend_deallocate(tensor)
+
+      _, :ok ->
+        :ok
     end)
   end
 
+  @doc """
+  Transforms the leading dimension in a tensor into a vectorized axis.
+
+  Nested vectorizations will "remove" the leading axis from the shape
+  and append it to the vectorized axes
+
+  {2, 3, 4}
+  |> vectorize(:first) # => {3, 4} vec [first: 2]
+  |> vectorize(:second) # => {4} vec [first: 2, second: 3]
+  """
```

Comment:
@josevalim ready for review again. I just need some pointers on what would be the ideal doc here.
Also, will we include a version that accepts a list of axes to vectorize at once as well?